decaf377-fmd/src/clue_key.rs

```diff
@@ -33,25 +34,47 @@ impl ClueKey {
     /// Fails if the bytes don't encode a valid clue key.
     #[allow(non_snake_case)]
     pub fn expand(&self) -> Result<ExpandedClueKey, Error> {
-        let root_pub_enc = decaf377::Encoding(self.0);
+        ExpandedClueKey::new(&self)
+    }
+}
+
+impl ExpandedClueKey {
+    pub fn new(clue_key: &ClueKey) -> Result<Self, Error> {
+        let root_pub_enc = decaf377::Encoding(clue_key.0);
         let root_pub = root_pub_enc
             .decompress()
             .map_err(|_| Error::InvalidAddress)?;
 
-        let Xs: [_; MAX_PRECISION] = (0..MAX_PRECISION)
+        Ok(ExpandedClueKey {
+            root_pub,
+            root_pub_enc,
+            subkeys: RefCell::new(Vec::new()),
+        })
+    }
+
+    /// Checks that the expanded clue key has at least `precision` subkeys
+    fn ensure_at_least(&self, precision: usize) -> Result<(), Error> {
+        if precision > MAX_PRECISION {
+            return Err(Error::PrecisionTooLarge(precision));
+        }
+
+        let current_precision = self.subkeys.try_borrow()?.len();
+
+        // The cached expansion is large enough to accomodate the specified precision.
+        if precision <= current_precision {
+            return Ok(());
+        }
+
+        let mut expanded_keys = (current_precision..precision)
             .into_iter()
-            .map(|i| hkd::derive_public(&root_pub, &root_pub_enc, i as u8))
-            .collect::<Vec<_>>()
-            // this conversion into an array will always succeed because we started with an iterator
-            // of length `MAX_PRECISION` and we're converting to an array of the same length
-            .try_into()
-            .expect("iterator of length `MAX_PRECISION`");
-
-        Ok(ExpandedClueKey { Xs })
+            .map(|i| hkd::derive_public(&self.root_pub, &self.root_pub_enc, i as u8))
+            .collect::<Vec<_>>();
+
+        self.subkeys.try_borrow_mut()?.append(&mut expanded_keys);
+
+        return Ok(());
```

Comment:
This function only returns an `Err` if there's a problem borrowing `self.subkeys`.  That can only happen if there's a conflicting borrow of `subkeys`, but we know that can't happen, because `subkeys` is only used internally to the `ExpandedClueKey`, and none of the borrows live longer than a single access.

Instead, I think it'd be cleaner to use the panicing `borrow()`/`borrow_mut()` functions (which we know won't panic), so that this function can be infallible.  Making this function infallible simplifies the downstream API:

* the new error conversions from the `RefCell` errors can be removed
* the `PrecisionTooLarge` error can stay as it is, only erroring if it's larger than `MAX_PRECISION`